### PR TITLE
Added HSM tile to interact with HSM

### PR DESCRIPTION
**Pull request by danTapps**
The events PR should be looked at first

### CODE REVIEW

File: nodes/hsm.js [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+      const options = { method: 'GET' };
+
+      try {
+        const response = await fetch(url, options);
+        if (response.status >= 400) {
+          node.status({ fill: 'red', shape: 'ring', text: 'response error' });
+          done(await response.text());
+          return;
+        }
+        node.currentHsm = output.response.hsm;
```
Comment (fblackburn1) [inline nodes/hsm.js]: I don't understand how it supposed to work, since output is not defined.
Is all node functionalities have been tested?

Comment (danTapps) [reply nodes/hsm.js]: That looks like I made a mistake here. That was supposed to be "response.hsm" as output get's defined after that.

File: nodes/hsm.js [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+          return 'armNight';
+        case 'off':
+        case 'disarm':
+        case 'disarmed':
+        case 'allDisarmed':
+        case 'alldisarmed':
+        case 3:
+          return 'disarm';
+        default:
+          return 'invalid';
```
Comment (fblackburn1) [inline nodes/hsm.js]: Where do you take all possibilities list?
How can I reproduce all these values?

**Comment by danTapps** [comment]
They are coming from the Homebridge plugin I have developed. These are a
collection of what I have seen while maintaining the Homebridge plugin and
working with users. Since they are only used on the input end, you can
easily pass them in as msg.command

**Comment by fblackburn1** [comment]
> They are coming from the Homebridge plugin I have developed. These are a collection of what I have seen while maintaining the Homebridge plugin and working with users. Since they are only used on the input end, you can easily pass them in as msg.command

Well ok. Any objection to remove number cases (ex: case: 0, case: 1, etc..) or they are real possibilities?

File: nodes/hsm.js [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+      }
+    }
+
+    node.on('input', async (msg, send, done) => {
+      let { command } = node;
+      if (msg.command !== undefined) {
+        command = msg.command;
+      }
+      if (!command) {
```
Comment (fblackburn1) [inline nodes/hsm.js]: How to get into this condition?
every node has a command by default

**Comment by danTapps** [comment]
Nope, no personal objection

File: nodes/hsm.js (hunk 3 above)
Comment (fblackburn1) [reply nodes/hsm.js]: Ah nvm, there is a bug, when you choose a command you cannot remove it. you need to close and reopen the node setting

**Comment by fblackburn1** [comment]
I think I would prefer to separate state node and setter node. 
I have done the PR for mode node #13 if you want to review it

To keep consistency, I would prefer to have the same logic for the hsm node
If you you agree and want to split it, I'll leave up to you
Otherwise, if you don't have time, just tell me and I will do it :wink:

**Comment by danTapps** [comment]
Hey, sorry for the late reply... I have my hands full at the moment, if you
want to, you can split it up. Or, if you rather have me to do it, I
probably will get to it next week.

**Comment by fblackburn1** [comment]
I'm still not able to figure out what is the difference between `armHome` and `armNight`.
From what I understand of the Hubitat interface, there are only `arm away` and `disable`
All other commands are custom and do not trigger anything
Do you have documentation or thread or announce or code to explain me?

**Comment by danTapps** [comment]
armNight and armHome are two different states in HSM, e.g:

armNight: arm the alarm with motion sensors active
armHome: arm the alarm with motion sensors inactive

They are really two different things.
These states are documented in the HSM documentation of Hubitat here:
https://docs.hubitat.com/index.php?title=Hubitat%C2%AE_Safety_Monitor

**Comment by fblackburn1** [comment]
> armNight and armHome are two different states in HSM, e.g: armNight: arm the alarm with motion sensors active armHome: arm the alarm with motion sensors inactive They are really two different things. These states are documented in the HSM documentation of Hubitat here: https://docs.hubitat.com/index.php?title=Hubitat%C2%AE_Safety_Monitor

Ah thank you, I've skip some documentation part :sweat_smile: 
My mistake was to have not set sensor on intrusion-night and intrusion-home. When you don't have sensor, even if you set armHome/armNight, the status is not changed

Everything makes sense now, thank you :+1:

**Comment by fblackburn1** [comment]
Well, I finished the code review and refactoring

I have only two questions about the current implementation

1. When receiving `hsmAlert` with value `cancel`/`cancelRuleAlerts` we fetch the hsm state. 
Why? Can we keep the currentHsm as it when we cancel alert as the state is not changed?
Ex:
Given HSM have state armHome
When I cancel alert
Then HSM state is still armHome
Are there some scenarios where cancelling alert also changes the hsm?

2. When alert is triggered, we receive an `hsmAlert` event with value (ex: `intrusion-home`). But we change the node status to this value, why?
If I restart the flow, the value will not be `intrusion-home`.
Maybe we need to have another node (ex: `hsm-alert`) to trigger alert events and keep the state if alert is triggered or not.
Another option is to send the event but do not change the node status text and change the status icon color when alert is triggered (red when alert, blue when cancelled)
I really like the last idea. What do you think? Do I miss something?

**Comment by fblackburn1** [comment]
My last commit explain how I see the thing, it may be clearer than words ;)

**Comment by danTapps** [comment]
Keep in mind, when I made the change I had a single setter/getter node, so
my approach was a little bit different.

1: you can probably survive without the refresh as the node would get
updated anyhow when the state changed

2: I would still change the node status text, maybe just append that an
alert was send with a time stamp but leave the original HSM state. The text
should reflect that an event was send, otherwise you will look for ages why
a certain sequence was triggered.

**Comment by fblackburn1** [comment]
:ok_hand: 
I have changed the node status text.

Can you try to include these nodes in your flows and/or do some manual tests before merge?
When I will have your Go, I will merge it

**Comment by fblackburn1** [comment]
@danTapps ping
(just want to be sure that you have been notified from my previous message, no pressure)

**Pull request merged**